# Requisições HTTP

O protocolo padrão de comunicação de dados na web é o HTTP.
Esse protocolo define uma série de padrões para que diferentes dispositivos possam se comunicar através da rede.

Quando queremos realizar uma requisição para um servidor utilizando o http precisamos enviar algumas informações como:
- Método da requisição (Method) <span style="color:red">*obrigatório*</span>
- Caminho (Path) <span style="color:red">*obrigatório*</span>
- Cabeçalhos (Headers) 


Este projeto possui um servidor bem simples, para ativa-lo abra um terminal e siga os passo:

```bash
cd ../server
npm run dev
```
Deve aparecer no terminal

Server is running on http://localhost:3000

Hora de testar nossa API!

In [1]:
#Importe a biblioteca requests ela será responsavel 
#por fazer as requisições ao servidor
import requests


r = requests.get('http://localhost:3000/')
print(r.text)

"Welcome to the server!"


Vamos explicar... estamos fazendo uma requisição http para o nosso servidor localhost
Podemos ver que a URL (caminho) é dividido em 4 partes
- protocolo (http)
- dominio/endereço (localhost)
- porta (3000)
- caminho (nesse caso é o **/**)

Ao realizarmos uma requisição (request) para o servidor, ele nos devolve uma resposta (response) nesse caso o texto "Welcome to the server!"

Esse objeto response nos tras algumas propriedades, como por exemplo o **status http**

In [7]:
r.status_code

200

O protocolo http nos trás um status code no header do response, esse status dependem da implementação da API que estamos acessando, contudo eles geralemente seguem alguns padrões...


<img src="static/status_code.png" style="text-center">

Nese caso tivemos um status 200 o que siginifica que a requisição ocorreu com sucesso!
Agora digamos que acessemos uma rota que não exista...

In [9]:
r = requests.get('http://localhost:3000/rota-que-nao-existe')
print(r.text, r.status_code)

Route not found 404


Recebemos um status 404 de Not Found, significa que não conseguimos acessar o **recurso** dessa rota

## Protocolos http

Nos exemplos acima apenas realizamos requisições utilizando o método **GET**, assim como o status code, temos diversos métodos https para realizar as requições de maneira correta

In [3]:
server = 'http://localhost:3000/'

In [20]:
r = requests.get(server + 'user/all')
print("Status Code: "+ str(r.status_code))
r.json()

[{'id': 1,
  'email': 'alice@email.com',
  'name': 'Alice',
  'type': None,
  'password': 'senha'},
 {'id': 2,
  'email': 'bob@email.com',
  'name': 'Bob',
  'type': None,
  'password': 'senha'},
 {'id': 3,
  'email': 'charlie@email.com',
  'name': 'Charlie',
  'type': None,
  'password': 'senha'},
 {'id': 4,
  'email': 'dave@email.com',
  'name': 'Dave',
  'type': None,
  'password': 'senha'},
 {'id': 5,
  'email': 'eve@email.com',
  'name': 'Eve',
  'type': None,
  'password': 'senha'},
 {'id': 6,
  'email': 'fulano@email',
  'name': 'Fulano',
  'type': None,
  'password': '123'}]

In [5]:
r = requests.get(server + 'user/html')
r.text

'<html><header><title>Lista de usuarios</title></header><body><h1>Lista de usuarios</h1><p>Nome: <b>Alice</b></p><li>Id: 1</li><li>Email: alice@email.com</li></br><p>Nome: <b>Bob</b></p><li>Id: 2</li><li>Email: bob@email.com</li></br><p>Nome: <b>Charlie</b></p><li>Id: 3</li><li>Email: charlie@email.com</li></br><p>Nome: <b>Fulano de Tal</b></p><li>Id: 4</li><li>Email: dave@email.com</li></br><p>Nome: <b>Eve</b></p><li>Id: 5</li><li>Email: eve@email.com</li></br></body></html>'

In [31]:
r = requests.get(server + 'user/count')
print("Status Code: "+ str(r.status_code))
r.json()

Status Code: 200


{'totalUsers': 6}

In [34]:
r = requests.get(server + 'user/4')
print("Status Code: "+ str(r.status_code))
r.json()

Status Code: 200


{'id': 4,
 'email': 'dave@email.com',
 'name': 'Dave',
 'type': None,
 'password': 'senha'}

In [32]:
new_user = {'name': 'Fulano', 'email': 'fulano@email', 'password': '123'}
r = requests.post(server + 'user', json=new_user)
print("Status Code: "+ str(r.status_code))
r.json()

Status Code: 400


{'error': 'Email já cadastrado'}

In [37]:
r = requests.put(server + 'user/4', json={'name': 'Fulano de Tal'})
print("Status Code: "+ str(r.status_code))
r.json()

Status Code: 203


{'message': 'Usuario atualizado com sucesso'}

In [35]:
r = requests.delete(server + 'user/6')
print("Status Code: "+ str(r.status_code))
r.json()

Status Code: 202


{'message': 'Deletado com sucesso'}

In [36]:
r = requests.get(server + 'user/all?limit=2&offset=2')
print("Status Code: "+ str(r.status_code))
r.json()

Status Code: 200


[{'id': 3,
  'email': 'charlie@email.com',
  'name': 'Charlie',
  'type': None,
  'password': 'senha'},
 {'id': 4,
  'email': 'dave@email.com',
  'name': 'Dave',
  'type': None,
  'password': 'senha'}]

In [41]:
r = requests.post(server + 'login', json={'email': 'admin@email.com', 'password': 'admin'})
token = r.json()['token']
r.json()

{'token': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImFkbWluQGVtYWlsLmNvbSIsInJvbGUiOiJhZG1pbiIsImlhdCI6MTcwODY0OTEzMywiZXhwIjoxNzA4NjQ5NDMzfQ.f7sqAabJzSnShTWKJtBG_NxZ77Kcqk2OgF7Dpzor_is'}

In [46]:
token_invalido = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImFkbWluQGVtYWlsLmNvbSIsInJvbGUiOiJhZG1pbiIsImlhdCI6MTcwODY0OTEzMywiZXhwIjoxNzA4NjQ5NDMzfQ.f7sqAabJzSnShTWKJtBG_NxZ77Kcqk2OgF7Dpzor_iA'
token_amigo = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6ImFkbWluQGVtYWlsLmNvbSIsInJvbGUiOiJhbWlnbyIsImlhdCI6MTcwODY0ODgxOSwiZXhwIjoxNzA4NjQ5MTE5fQ.NpzMRBTCFkJPCGLcZrV1hMcgZ4w43ug6LvOj3A6OGac'
r = requests.get(server + 'admin', headers={'authorization': 'Barear '+token})
r.json()

{'message': 'Meus dados sigilosos'}